x/iro/simulation/operations.go

```diff
@@ -0,0 +1,153 @@
+		}
+
+		// Randomly select a plan
+		plan := plans[r.Intn(len(plans))]
+		curve := plan.BondingCurve
+
+		// Test different token amounts
+		testAmounts := []math.Int{
+			math.NewInt(1).MulRaw(1e17),      // 0.1 token
+			math.NewInt(1).MulRaw(1e18),      // 1 token
+			math.NewInt(100).MulRaw(1e18),    // 100 tokens
+			math.NewInt(1000).MulRaw(1e18),   // 1000 tokens
+			math.NewInt(10000).MulRaw(1e18),  // 10000 tokens
+			math.NewInt(100000).MulRaw(1e18), // 100000 tokens
+		}
+
+		// prepare base error with curve context
+		totalAll := types.ScaleFromBase(plan.TotalAllocation.Amount, types.DYMDecimals)
+		soldAmt := types.ScaleFromBase(plan.SoldAmt, types.DYMDecimals)
+		targetRaise := types.ScaleFromBase(curve.Cost(math.ZeroInt(), plan.TotalAllocation.Amount), types.DYMDecimals)
+		curveDesc := fmt.Sprintf("total supply: %s, sold amount: %s, target raise: %s, bonding curve: %s", totalAll.String(), soldAmt.String(), targetRaise.String(), curve.Stringify())
+
+		if curve.M.IsZero() {
+			err := fmt.Errorf("%s: M is zero", curveDesc)
+			return simtypes.NoOpMsg(types.ModuleName, "TestBondingCurve", err.Error()), nil, err
+		}
+
+		var results []string
+		managed := false
+		lastCost := math.ZeroInt()
+		for _, amount := range testAmounts {
+			scaledAmount := types.ScaleFromBase(amount, types.DYMDecimals)
+			// Calculate cost for buying tokens
+			cost := curve.Cost(plan.SoldAmt, amount.Add(plan.SoldAmt))
+			if !cost.IsPositive() {
+				continue
+			}
+
+			// Calculate tokens for exact DYM spend
+			tokens, err := curve.TokensForExactDYM(plan.SoldAmt, cost)
+			if err != nil {
+				err = fmt.Errorf("%s: buy amount: %s,tokens for exact DYM spend: %w", curveDesc, scaledAmount.String(), err)
+				return simtypes.NoOpMsg(types.ModuleName, "TestBondingCurve", err.Error()), nil, err
+			}
+
+			// validate tokens are approximately the same as the amount
+			if err := approxEqualInt(amount, tokens); err != nil {
+				err = fmt.Errorf("%s: buy amount: %s, tokens not equal to amount: %w", curveDesc, scaledAmount.String(), err)
+				return simtypes.NoOpMsg(types.ModuleName, "TestBondingCurve", err.Error()), nil, err
+			}
+
+			if cost.LTE(lastCost) {
+				err = fmt.Errorf("%s: cost not increasing: %s <= %s", curveDesc, cost.String(), lastCost.String())
+				return simtypes.NoOpMsg(types.ModuleName, "TestBondingCurve", err.Error()), nil, err
+			}
+
+			lastCost = cost
+			managed = true
+
+			results = append(results, fmt.Sprintf(
+				"Amount: %s tokens, Cost: %s DYM, TokensForExactDYM: %s tokens",
+				amount.String(), cost.String(), tokens.String(),
+			))
+		}
+
+		if !managed {
+			err := fmt.Errorf("%s: no valid cost found for any of the test amounts", curveDesc)
+			return simtypes.NoOpMsg(types.ModuleName, "TestBondingCurve", err.Error()), nil, err
+		}
+
+		return simtypes.NewOperationMsg(&types.MsgBuy{}, true, fmt.Sprintf("%s Results:\n%s", curveDesc, results), cdc), nil, nil
+	}
+}
```

Comment:
This looks like an invariant or randomized test, it's not a simulated operation?
I'm not sure I understand the approach here. Seems like a very weird use of sdk simulation

Reply:
used the framework as simulation /  fuzzing framework
It's too much effort to prepare all the preconditions for actual Buy/Sell Msgs

used the simulation framework to generate different curves, at simulate different cost functions

Reply:
Got you. There are much better tool for that like go stdlib fuzz or rapid FYI